digitalocean/reservedipv6/resource_reserved_ipv6.go

```diff
@@ -0,0 +1,226 @@
+package reservedipv6
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/digitalocean/godo"
+	"github.com/digitalocean/terraform-provider-digitalocean/digitalocean/config"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+func ResourceDigitalOceanReservedIPV6() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceDigitalOceanReservedIPV6Create,
+		UpdateContext: resourceDigitalOceanReservedIPV6Update,
+		ReadContext:   resourceDigitalOceanReservedIPV6Read,
+		DeleteContext: resourceDigitalOceanReservedIPV6Delete,
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceDigitalOceanReservedIPV6Import,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"region_slug": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				StateFunc: func(val interface{}) string {
+					// DO API V2 region slug is always lowercase
+					return strings.ToLower(val.(string))
+				},
+			},
+			"urn": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "the uniform resource name for the reserved ipv6",
+			},
+			"ip": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ValidateFunc: validation.IsIPv6Address,
+			},
+			"droplet_id": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceDigitalOceanReservedIPV6Create(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*config.CombinedConfig).GodoClient()
+
+	log.Printf("[INFO] Creating a reserved IPv6 in a region")
+	regionOpts := &godo.ReservedIPV6CreateRequest{
+		Region: d.Get("region_slug").(string),
+	}
+
+	log.Printf("[DEBUG] Reserved IPv6 create: %#v", regionOpts)
+	reservedIP, _, err := client.ReservedIPV6s.Create(context.Background(), regionOpts)
+	if err != nil {
+		return diag.Errorf("Error creating reserved IP: %s", err)
+	}
+
+	d.SetId(reservedIP.IP)
+
```

Comment:
This should assign the droplet to the reserved IPv6.
Config example:
```
resource "digitalocean_droplet" "foobar12" {
  image  = "ubuntu-22-04-x64"
  name   = "foobar-12"
  region = "nyc3"
  size   = "s-1vcpu-1gb"
  ipv6   = true
}

resource "digitalocean_reserved_ipv6" "rfoobar12" {
  droplet_id  = digitalocean_droplet.foobar12.id
  region_slug = digitalocean_droplet.foobar12.region
}
```

```suggestion

	if v, ok := d.GetOk("droplet_id"); ok {
		log.Printf("[INFO] Assigning the reserved IP to the Droplet %d", v.(int))
		action, _, err := client.ReservedIPV6Actions.Assign(context.Background(), d.Id(), v.(int))
		if err != nil {
			return diag.Errorf(
				"Error Assigning reserved IPv6 (%s) to the Droplet: %s", d.Id(), err)
		}

		_, unassignedErr := waitForReservedIPV6Ready(ctx, d, "completed", []string{"new", "in-progress"}, "status", meta, action.ID)
		if unassignedErr != nil {
			return diag.Errorf(
				"Error waiting for reserved IPv6 (%s) to be assigned: %s", d.Id(), unassignedErr)
		}
	}

```

Reply:
@loosla  for assignment, we have different terraform resource `digitalocean_reserved_ipv6_assignment` , reservedIPv6 doesn't support Reserve and Assign in one API as compared to reservedIP v4, so keeping this behaviour in sync with API level

Reply:
So, for reserve and assign in tf file, customers can use this 
```
resource "digitalocean_reserved_ipv6" "foobar" {
  region_slug = "nyc3"
}

resource "digitalocean_droplet" "foobar" {
  count  = 1
  name   = "tf-acc-test-${count.index}"
  size   = "s-1vcpu-1gb"
  image  = "ubuntu-22-04-x64"
  region = "nyc3"
  ipv6   = true
}

resource "digitalocean_reserved_ipv6_assignment" "foobar" {
  ip         = digitalocean_reserved_ipv6.foobar.ip
  droplet_id = digitalocean_droplet.foobar.0.id
}

```

Reply:
:+1: Thanks for explanation.

What is the purpose of having `droplet_id` in `ResourceDigitalOceanReservedIPV6` if users can't use it to set or update the value?

I think this is going to be a poor user experience that will lead to TF issues and support tickets.

If users must use `digitalocean_reserved_ip_assignment` should `droplet_id` be removed from `ResourceDigitalOceanReservedIPV6` to avoid confusion?

Reply:
Umm, once assigned, a reservedIPv6 can have a droplet_id but in the case of create I was going with in sync with APIs

Reply:
Thanks Andrew, figured that out